memtable/inlineskiplist.h

```diff
@@ -838,8 +859,23 @@ bool InlineSkipList<Comparator>::Insert(const char* key, Splice* splice,
         assert(splice->prev_[i] == head_ ||
                compare_(splice->prev_[i]->Key(), x->Key()) < 0);
         x->NoBarrier_SetNext(i, splice->next_[i]);
+        if (i == 0) {
+          x->NoBarrier_SetPrev(splice->prev_[0]);
+        }
+
         if (splice->prev_[i]->CASNext(i, splice->next_[i], x)) {
           // success
+          if (i == 0) {
+            if (splice->next_[0] != nullptr) {
+              Node* prev = splice->prev_[0];
+              Node* next = splice->next_[0];
+              while (!next->CASPrev(&prev, x)) {
```

Comment:
Here you seems to assume if `CASPrev` fail, there is a node inserted before `x` (say it is node `y`), and so you can retry CAS to insert between `next` and `y`. However, the node `y` can be after (larger than) `x`. Then it is incorrect to insert `x` between `y` and `next`. You need to follow `next->prev` to find the new insert position before retry CAS. And you forget to update `x->prev` when CAS fail.

And note that it is possible `splice->next[0]` haven't even linked on the reverse linked-list at this point, since we use forward linked-list to calculate `splice`.

To fix the above issues, one approach I can think of is
```
void InsertForPrev(x, key_decoded, splice) {
  // Note the splice is only used for the first attempt. Since we always insert to 
  // reverse linked-list before insert to the forward linked-list, nodes on the splice is
  // guaranteed to be on the reverse linked-list.
  Node* prev = splice->prev_[0];
  Node* next = splice->next_[0];
  if (next == nullptr) {
    // We need to add a tail node, otherwise we don't know where to insert to in this case.
    next = tail_;
  }
  x->NoBarrier_SetPrev(prev);
  while (!next->CASPrev(prev, next) {
    // use reverse linked-list to search for new insert position.
    prev = next->Prev();
    while (!KeyIsAfterNode(key_decoded, prev)) {
      next = prev;
      prev = next->Prev();
    }
    x->NoBarrier_SetPrev(prev);
  }
}

bool Insert(...) {
  ...
  // Insert to reverse linked-list before inserting to forward linked-list.
  InsertForPrev();
  // Insert for next.
  for (int i = 0; i < height; i++) {
    ...
  }
  ...
}
```